### ISSUE TITLE
com.twitter.zipkin.gen.Span vs Brave.Span : how to link REST span with programatic span , 

### ISSUE THREAD

**Issue by LasneF** [openzipkin/brave#351]
i am using a BraveServletFilter properly configured. Then i want to create span manualy and attach them to the current Span . 
so the natural way would have be to do this Span child = tracer.newChild( ??? )
but how to get the current span of my rest call  ? 
one way would be to use
com.twitter.zipkin.gen.Span localSpan = brave.serverSpanThreadBinder().getCurrentServerSpan().getSpan();
Span child = tracer.newChild(localSpan ) 

but unfortunately here Brave leaks an internal com.twitter.zipkin.gen.Span and not a brave.Span type 

then i get to complex code to convert (not done yet ) i was wondering is there a better way 

should Brave stop leaking internal twitter package class.

**Comment by adriancole** [comment]
you need to create the Brave instance with TracerAdapter, then you can use the instructions mentioned here https://github.com/openzipkin/brave/tree/master/brave#server-spans

**Comment by LasneF** [comment]
oh i may mix peer and apple in term of version, my project is new so would like to use the most accurate way not mixing version 

Tracer tracer = Tracer.newBuilder()
            .traceId128Bit(traceId128Bit)
            .reporter(reporter)
            .sampler(sampler)
            .build();
            
 Brave brave3 = TracerAdapter.newBrave(tracer)          
 BraveServletFilter sevBraveServletFilter = BraveServletFilter.create(brave3)           
 and install it  my servlet filter into my container 

NB  : a put Brave and Tracer as singleton to get access to them in my end points

then in my rest end point should i do this  ? 
brave3Span = brave3.localSpanThreadBinder().getCurrentLocalSpan();
brave4Span = TracerAdapter.toSpan(tracer, brave3Span);
tracer.newChild(brave4Span.context()).name("My Children")

there is no more convenient way ?
or may be i did not get the full picture :(

**Comment by adriancole** [comment]
>
> oh i may mix peer and apple in term of version, my project is new so would
> like to use the most accurate way not mixing version
>
agreed. what you are seeing is a work in progress to completely transition
everything in Brave to a new set up apis. It will complete, just it isn't
yet. here's the latest status

https://github.com/openzipkin/brave/issues/180#issuecomment-275033564


> Tracer tracer = Tracer.newBuilder()
> .traceId128Bit(traceId128Bit)
> .reporter(reporter)
> .sampler(sampler)
> .build();
>
> Brave brave3 = TracerAdapter.newBrave(tracer)
> BraveServletFilter sevBraveServletFilter = BraveServletFilter.create(
> brave3)
> and install it my servlet filter into my container
>
> NB : a put Brave and Tracer as singleton to get access to them in my end
> points
>
yep

> then in my rest end point should i do this ?
> brave3Span = brave3.localSpanThreadBinder().getCurrentLocalSpan();
> brave4Span = TracerAdapter.toSpan(tracer, brave3Span);
> tracer.newChild(brave4Span.context()).name("My Children")
>
> there is no more convenient way ?
>
there will certainly be a more convenient way, in a matter of weeks
assuming no-one else helps with the work involved.

> or may be i did not get the full picture :(
>
nope you got it right. the end goal of the new brave apis is such that
green field projects do not have to use the old packages. Tactically, the
highest value was getting a core api in place and making sure that the api
could backport existing instrumentation. That was temporarily optimizing
existing users over new users, and was a *ton* of work to make happen.

So, basically the idea is to use the bridge approach for now, and as soon
as possible, you can tear it out with a clean copy. While
https://github.com/openzipkin/brave/pull/350 is not in a form that's worth
testing, I'll ping you on it when ready, so that you can help ensure the
new apis are in fact as easy to use as you'd expect.

**Comment by LasneF** [comment]
Thx you for all your clarification , all is cristal clear ,  thx for all your efforts and your explanation
